caper/resource_analysis.py

```diff
@@ -0,0 +1,338 @@
+            plot_pp.close()
+
+        return result
+
+    def analyze_task(
+        self,
+        task_name,
+        in_file_vars=None,
+        reduce_in_file_vars=DEFAULT_REDUCE_IN_FILE_VARS,
+        target_resources=DEFAULT_TARGET_RESOURCES,
+        plot_pp=None,
+    ):
+        """Does resource analysis on a task.
+        To use a general solver for y = f(x),
+        convert task's raw resouce data into (x_matrix, y_vector) form.
+
+        Input file sizes are converted into one single matrix.
+        Such x_matrix should look like:
+            [
+                [IN_FILE1_SIZE1, IN_FILE2_SIZE1, ...],
+                [IN_FILE1_SIZE2, IN_FILE2_SIZE2, ...],
+                ...
+            ]
+        If reduce_in_file_vars is defined (e.g. sum).
+        Such x_matrix will be reduced to a vector.
+            [
+                [SUM_ALL_IN_FILES_SIZE1],
+                [SUM_ALL_IN_FILES_SIZE2],
+                ...
+            ]
+
+        Each resource metric in `target_resources` becomes a y_vec.
+        For example, there are two kinds of default resource metrics.
+        i.e. max_mem and max_disk.
+        Therefore, y_vec_1 should look like:
+            [
+                STATS_MAX_MEM1,
+                STATS_MAX_MEM2,
+                ...
+            ]
+        y_vec_2 should look like:
+            [
+                STATS_MAX_DISK1,
+                STATS_MAX_DISK2,
+                ...
+            ]
+        They are two separate problems.
+
+        Args:
+            in_file_vars:
+                List of input file vars. [IN_FILE_VAR1, ...].
+                Matching input file will be included in x_matrix of analysis.
+                See `input_file_sizes` in CromwellMetadata.gcp_monitor.__doc__ for details.
+                If None or False, then all input files will be used to make x_matrix.
+            reduce_in_file_vars:
+                Python function to reduce x_vector. e.g. sum, max.
+                If None, then no reduction of vector x.
+                If you don't reduce x, make sure that you have enough number of data (>=dim(x)).
+                Some solver can fail due to lack of data. i.e. dim(problem, solver) > dim(data).
+                For example, even a linear solver will fail with only one dataset.
+            target_resources:
+                Keys (in dot notation) to make vector y. e.g. ('stats.max.mem', 'stats.max.disk').
+                See CromwellMetadata.gcp_monitor.__doc__ to find available keys.
+                One vector y for each key, which means one solving for each key.
+            plot_pp:
+                Matplotlib's PDF backend PdfPages to write plots on multiple pages
+                (task per page).
+
+        Returns:
+            coeffs:
+                Analysis result.
+                e.g. (coeffs, intercept) for linear regression.
+            Plot object:
+                Matplotlib plot object.
+        """
+        import numpy as np
```

Comment:
Would prefer to see the import at the top of the file.

Reply:
Moved to the top.